src/macaron/policy_engine/souffle.py

```diff
@@ -0,0 +1,179 @@
+# Copyright (c) 2023 - 2023, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""
+Wrapper classes for invoking souffle by subprocess and getting the resulting tables.
+
+Implements a context manager to create and clean up temporary directories.
+"""
+
+import csv
+import glob
+import logging
+import os
+import shutil
+import subprocess  # nosec B404
+import tempfile
+from types import TracebackType
+from typing import Optional
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+class SouffleError(Exception):
+    """Occurs when the souffle program contains errors, or there is an error invoking souffle."""
+
+    # TODO: Use generic Macaron error class
+
+    def __init__(
+        self, command: Optional[list[str] | str] = None, message: str = "An error occurred with calling Souffle."
+    ):
+        self.message = message
+        self.command = command
+        super().__init__(self.message)
+
+
+class SouffleWrapper:
+    """Wrapper class for managing the temporary working directory of the souffle interpreter.
+
+    Example
+    -------
+    with SouffleWrapper(fact_dir="facts", output_dir="output") as sfl:
+        text = "<souffle program>"
+        result = sfl.interpret_text(text)
+        assert result == {"path": [["1", "2"], ["1", "3"], ["2", "3"]]}
+
+    """
+
+    _souffle: str
+    output_dir: str
+    include_dir: str
+    fact_dir: str
+    library_dir: str
+    souffle_stdout: Optional[str]
+    souffle_stderr: Optional[str]
+
+    TEMP_SOURCEFILE_NAME = "source.dl"
+
+    def __init__(
```

Comment:
Could you add some documentation on what the parameters are? I think the ones that might be confusing are `include_dir` vs `library_dir`.

Reply:
Fixed here https://github.com/oracle-samples/macaron/pull/46/commits/4e61112595339f7871796e707f604cc1ff1b5711